src/gallium/drivers/lima/lima_draw.c

```diff
@@ -998,6 +1015,18 @@ lima_flush(struct pipe_context *pctx, struct pipe_fence_handle **fence,
    lima_submit_add_bo(ctx->pp_submit, ctx->share_buffer, LIMA_SUBMIT_BO_READ);
    lima_submit_add_bo(ctx->pp_submit, ctx->pp_buffer,
                       LIMA_SUBMIT_BO_READ|LIMA_SUBMIT_BO_WRITE);
+   if (ctx->tex_stateobj.num_samplers) {
+      struct lima_texture_stateobj *lima_tex = &ctx->tex_stateobj;
+      lima_submit_add_bo(ctx->pp_submit, ctx->tex_descs, LIMA_SUBMIT_BO_READ);
+      for (int i = 0; i < ctx->tex_stateobj.num_textures; i++) {
+         struct lima_sampler_view *texture = lima_sampler_view(lima_tex->textures[i]);
+         struct pipe_resource *prsc = texture->base.texture;
+         struct lima_resource *lima_res = lima_resource(prsc);
+
+         lima_submit_add_bo(ctx->pp_submit, lima_res->bo, LIMA_SUBMIT_BO_READ);
```

Comment:
This should be done inside draw_vbo which will capture the textures needed when call glDraw*.